Nexus attributes verbesserung

cstamas opened the pull request:
A massive one:
- opening up for other serializators, not being chained to XStream (as today)
- slimming down AttributeStorage iface, because since Nexus 1.4 there was no need to persist whole StorageItem, the attributes was enough (it was a leftover since Proximity)
- StorageItem cleanup, before it was: a bunch of transient fields, a bunch of persisted fields and one persisted map. Now, an item is: transient map (context, as before), persisted map (attributes, and everything else, the actual data)
- upgrade path provided: if attributes detected in old place, "legacy" AttributeStore kicks in and Nexus seamlessly upgrades attribues to new ones (as Nexus is used).
- current attribute storage is LS attribute storage (attributes as below storage/repo-id/.nexus/attributes) and have new, cleaner format, they are just a Map (series of key-val pairs).

In previous pull https://github.com/sonatype/nexus/pull/93 Marshaller iface is already provided. There, only the XStreamMarshaller was viable (jacksons were unable to cope with Item classes). Here, we have them all: XstreamMarshaller, but JacksonJSON and JacksonXML worx too! As I measured, JacksonJSON is a bit better over Xstream (about 10%-ish), but this is CPU only, since IO is clearly the same (no much size differences).

And once Jackson enters core, nothing will stop it to come all way to the REST API ;)
Just watch! :beer:

cstamas committed:
Nexus Attributes Verbesserung!

In short: fixing the 4+ year old problem how attributes were handled in Nexus.

cstamas committed:
Added "transitioning" AttributeStorage that does upgrade.
Simplified and concentrated attribute decisions into one place: DelegatingAttributeStorage

cstamas committed:
V!

Added header and javadoc.

cstamas committed:
One failing UT fixed:

It was doing assertion against size of "listing of root directory", but it is no more 4 but 5, since ".nexus" popped into game with LS AttributeStorage.

cstamas committed:
Adding missing header.

cstamas committed:
Attributes implementation simplified, and ensured backward compat.

Attributes are not a Map anymore. Adapter provided for old code
that still mutates item attributes over item.getAttributes() deprecated
method.

cstamas committed:
Bumping Surefire to latest.

cstamas committed:
Nexus nexus state event borked attribute manager.

cstamas committed:
Applying annos at proper places

cstamas committed:
Adding UT and IT tests covering legacy Attribute store, new attribute store and the "transitioning upgrade"

cstamas commented:
Scheduled IT voting:
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/69/

But, already realized this vote will not pass. Just like the previous feature job building this branch
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/62/

There are 4 existing (not newly added in this pull req) that I did not yet update. They all do assertions against attributes and expect them in old place. They will still fail. Fix coming.

nabcos commented:
Build failed hard before ITs:

Nexus4660AttributeStorageFSLSTransitioningUpgrade.java:[34,47] package com.sun.tools.internal.ws.processor.util does not exist

cstamas committed:
Removing unused imports from IT class

cstamas commented:
Fixed, new job started (still 4 IT expected to fail):
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/70/

nabcos commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/attributes/internal/AttributesImpl.java:
```diff
@@ -0,0 +1,350 @@
+ * @since 1.10.0
+ */
+public class AttributesImpl
+    implements Attributes
```
Do we really need another naming convention in the nexus codebase?

velo replied:
#true

nabcos commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/attributes/internal/AttributesImpl.java:
```diff
@@ -0,0 +1,350 @@
+
+    protected String getKeyForAttribute( final String attributeName )
+    {
+        // TODO: Jackson will use the key as XML element!
```
I don't get this TODO, is it still relevant/in the right place?

velo commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/item/AbstractStorageItem.java:
```diff
@@ -411,39 +494,41 @@ public abstract class AbstractStorageItem
-            throw new IllegalArgumentException( "Cannot overlay storage item of class " + item.getClass().getName()
-                + " onto this item of class " + this.getClass().getName() );
-        }
+        // this is noop, is not used in core at all
```
Why not remove this then?

nabcos commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/item/AbstractStorageItem.java:
```diff
@@ -19,118 +19,209 @@
+     * @param path       the path
+     * @param readable   the readable
+     * @param writable   the writable
      */
```
pretty useless javadoc here, IMO better remove it

peterlynch replied:
or fix it

nabcos commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/item/AbstractStorageItem.java:
```diff
@@ -19,118 +19,209 @@
+     * @param virtual  the virtual
      * @param readable the readable
      * @param writable the writable
      */
```
useless javadoc

nabcos commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/item/AbstractStorageItem.java:
```diff
@@ -315,14 +402,10 @@ public abstract class AbstractStorageItem
         return ItemPathUtils.getParentPath( getPath() );
     }
 
+    @Deprecated
```
There should be a note in javadoc why it's deprecated.

peterlynch replied:
+1 always

velo commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/AbstractAttributeStorage.java:
```diff
@@ -19,35 +19,16 @@
  * @author cstamas
+ * @since 1.10.0
  */
 public abstract class AbstractAttributeStorage
```
Why this is need since nothing is implemented?

cstamas replied:
See here:
https://github.com/sonatype/nexus/pull/79#r236711

nabcos commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/item/AbstractStorageItem.java:
```diff
@@ -337,72 +420,72 @@ public abstract class AbstractStorageItem
      * Sets the stored locally.
-     * 
+     *
      * @param storedLocally the new stored locally
```
IDE was probably stoned locally when it was setting this javadoc :o)

nabcos commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/item/AbstractStorageItem.java:
```diff
@@ -411,39 +494,41 @@ public abstract class AbstractStorageItem
-        }
+        // this is noop, is not used in core at all
+
+        // if ( isOverlayable( item ) )
```
dead code should be removed, if necessary note what the dead code did before.

velo commented:
+1

nabcos commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/item/StorageItem.java:
```diff
@@ -175,10 +176,18 @@ public interface StorageItem
      * @return the attributes
+     * @deprecated Use {@link #getRepositoryItemAttributes()} instead!
      */
     Map<String, String> getAttributes();
```
`@Deprecated` annotation missing
http://docs.oracle.com/javase/1.5.0/docs/guide/javadoc/deprecation/deprecation.html#annotation

nabcos commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/item/StorageItem.java:
```diff
@@ -189,10 +198,20 @@ public interface StorageItem
      * @param item the item
+     * @deprecated Whoever used this outside Nexus core is doomed.
      */
     void overlay( StorageItem item );
```
`@Deprecated` missing

cstamas replied:
Actually, based on the link above (deprecation annotation use), this method should be "retracted" really.

nabcos commented on nexus/pom.xml:
```diff
@@ -187,7 +187,7 @@
     <spice-timeline.version>2.1-SNAPSHOT</spice-timeline.version>
     <spice-utils.version>1.0.0</spice-utils.version>
-    <surefire.version>2.9</surefire.version>
+    <surefire.version>2.10</surefire.version>
```
Unrelated change? ;)

cstamas replied:
Um, yes and no:
https://issues.sonatype.org/browse/NXCM-3481

but i did that after I hit the OOM (that mentioned in referenced Surefire issue) in a local full build.

nabcos commented:
+1

cstamas committed:
Reverted back to "plexus naming"

cstamas committed:
Applied remarks, but also removed the newly added StorageItem#overlayAttributes()
since they are redundant.

cstamas commented:
Good, CI confirmed: the 4 expected IT failed (they expect attributes in old place)
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/70/

adreghiciu commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/attributes/AttributeStorage.java:
```diff
@@ -32,25 +30,26 @@ import org.sonatype.nexus.proxy.storage.local.LocalRepositoryStorage;
 public interface AttributeStorage
```
Use this opportunity to fix the java doc of this class by actually explaining what it is

cstamas replied:
Latest change covers that, but not only that, but also deprecated a lot of piled up oldie methods (read: cleanup)

adreghiciu commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/DelegatingAttributeStorage.java:
```diff
@@ -18,73 +18,84 @@
- * "fallback" to some other instance. Usable for scenarios where "transitioning" (smooth upgrade for example) is to be
- * used, the "main" attribute storage would be "upgraded" from "legacy" attribute storage as the attributes are
- * requested over the time from this instance.
+ * Simple wrapping AttributeStorage that delegates only when needed.
```
JavaDoc should explain when the delegation is performed

adreghiciu commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/DelegatingAttributeStorage.java:
```diff
@@ -18,73 +18,84 @@
+    {
+        Boolean isMetadataMaintained = uid.getAttributeValue( IsMetadataMaintainedAttribute.class );
+
+        if ( isMetadataMaintained != null )
```
The lines bellow can eb replaced by:

return isMetadataMaintained == null || isMetadataMaintained;

adreghiciu commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/TransitioningAttributeStorage.java:
```diff
@@ -0,0 +1,96 @@
+
+        if ( fallbackAttributeStorage != null )
+        {
+            fallbackAttributeStorage.deleteAttributes( uid );
```
Delete is not needed as the attributes are already deleted on get(). It is safer lets say but is not needed.

cstamas replied:
I like to be on safe side. Also, in some circumstances, without delete here, it would be impossible to distinguish situation like "transitioning is done": all the attributes were either upgraded or just deleted.

adreghiciu commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/TransitioningAttributeStorage.java:
```diff
@@ -0,0 +1,96 @@
+    public boolean deleteAttributes( final RepositoryItemUid uid )
+    {
+        return mainAttributeStorage.deleteAttributes( uid )
+            || ( fallbackAttributeStorage != null && fallbackAttributeStorage.deleteAttributes( uid ) );
```
Delete is not needed on fallback as the attributes are already deleted on get(). It is safer lets say but is not needed.

cstamas replied:
I like to be on safe side. Also, in some circumstances, without delete here, it would be impossible to distinguish situation like "transitioning is done": all the attributes were either upgraded or just deleted.

adreghiciu commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/LegacyFSAttributeStorage.java:
```diff
@@ -0,0 +1,370 @@
+ * Legacy AttributeStorage implementation that uses it's own FS storage to store attributes, by persisting StorageItem
+ * as whole using XStream. This is the "old" default storage, used in all Nexuses up to version 1.10.0. Note: this
+ * component is ReadOnly and should be used to perform transitioning upgrade from older Nexus instances. Only
+ * {@link #deleteAttributes(RepositoryItemUid)} and {@link #getAttributes(RepositoryItemUid)} will do anyting, while
```
typo "anyting"

adreghiciu commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/LegacyFSAttributeStorage.java:
```diff
@@ -0,0 +1,370 @@
+@Typed( AttributeStorage.class )
+@Named( "legacy" )
+@Singleton
+public class LegacyFSAttributeStorage
```
What is the diff between this and DefaultFSAttributeStorage?

cstamas replied:
Legacy is what master have now: wired in XStream, and handles StorageItem persistence. The DefaultFSAttributeStorage is using Marshaller and deals with Attributes. Btw, this change was introduced with https://github.com/sonatype/nexus/pull/93  This pull request just brings back the "legacy" to support transitioning upgrade.

Frankly, DefaultFSAttributeStorage is not needed anymore, but is left as a "safe fallback" if we decide to not move attributes into Repository storage for any reason.

adreghiciu commented:
You could fix the AttributesHandler java doc on class/methods

cstamas committed:
Added locking to transitioning AttributeStorage to properly guard eventual upgrades.

cstamas committed:
Applying pull req comments, but also deprecating a bunch of methods appeared in time
and created two focused for them instead.

bdemers commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/attributes/AttributeStorage.java:
```diff
@@ -32,25 +30,26 @@ import org.sonatype.nexus.proxy.storage.local.LocalRepositoryStorage;
+     * @param attributes the attributes to store
      */
-    void putAttribute( StorageItem item );
+    void putAttributes( RepositoryItemUid uid, Attributes attributes );
```
api change, this at least needs @since

bdemers commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/attributes/AttributeStorage.java:
```diff
@@ -32,25 +30,26 @@ import org.sonatype.nexus.proxy.storage.local.LocalRepositoryStorage;
+     * @param uid the key for which attributes needs to be fetched
      * @return the attributes
      */
-    AbstractStorageItem getAttributes( RepositoryItemUid uid );
```
API signature change

bdemers commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/item/AbstractStorageItem.java:
```diff
@@ -19,133 +19,204 @@
+    }
+
+    /**
+     * This method should be called ONLY when you load up a _legacy_ attribute using _legacy_ attribute store!
```
What is the _legacy_ store ?   Does this commit assume we are moving the attribute store to the repo storage folder ?

cstamas replied:
Not this commit but the whole pull request does, just like described in pull request description. 

I left this as "open question" (FS AttributeStorage is still there), but IMO doing Attributes format change (needed to open up for other than XStream and improving the serialization speed in general) without the "attributes relocation under repository storage" (the one we already did once, and then backed out from it as we were unsure about upgrade path) would not make sense. Move is needed anyway to have distinct storages for "old" and "new" attributes (format change happens), so if we move, why not move to the place where we almost did once?

Also, this was discussed/mentioned multiple times (but always postponed or kicked out from iterations), see issues like NXCM-3492, NXCM-2743 etc.

bdemers replied:
OK, my mistake, I didn't understand the first time I read this.

We should chat about this.  Do we want to move the attribute storage?

peterlynch commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/attributes/AttributesHandler.java:
```diff
@@ -27,89 +27,121 @@ import org.sonatype.nexus.proxy.item.StorageItem;
+     * Sets the attribute storage used by this instance.
+     *
+     * @param attributeStorage
+     * @deprecated For UT uses only!
```
@VisibleForTesting ?

peterlynch commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/DefaultAttributesHandler.java:
```diff
@@ -122,12 +125,12 @@ public class DefaultAttributesHandler
-                                     @Named( "fs" ) AttributeStorage attributeStorage,
+                                     @Named( "transitioning" ) AttributeStorage attributeStorage,
                                      List<StorageItemInspector> itemInspectorList,
                                      List<StorageFileItemInspector> fileItemInspectorList )
```
curious why Preconditions.checkNotNull() not used here.

peterlynch commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/DefaultAttributesHandler.java:
```diff
@@ -136,27 +139,29 @@ public class DefaultAttributesHandler
-     * 
+     *
      * @param attributeStorage the new attribute storage
+     * @deprecated For UT use only!
```
@VisibleForTesting ?

peterlynch commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/DefaultAttributesHandler.java:
```diff
@@ -362,19 +379,93 @@ public class DefaultAttributesHandler
         return doTouch;
     }
 
+    // == Below is junkyard, deprecated methods
```
don't we need @deprecated and reason why?

cstamas replied:
I could add as "inherited docs". This is the default and hence, the _only_ AttributeHandler implementation, and the interface already contains the explanation and pointer to what use instead.

cstamas committed:
Fixing ITs: 3 out of affected by attribute storage location change.

The nexus2692 IT is intact, and will be the only one failing, but IMO that IT is
too dirty and does illegal (?) things, and should be reimplemented --- tests evict
proxied items against hosted reposes for example, etc.

cstamas commented:
Rebase happened against latest master.

cstamas committed:
Added a nullcheck to make some ITs work

cstamas committed:
Fixed the last IT failing after attributes change.

This IT did some extensive low-level manipulation with attributes to
test the "evict unused cache items" task. The fix was twofold:

* fixed how attributes are "manipulated" (keeping them in old format, since
  during IT run they will be upgraded anyway)
* stopped doing separate checks against attributes only, since the task
  removes item. This does not lessens the value of this IT, since
  nexus takes care about attribute removal when item is removed, and
  this IT is anyway not about that.

cstamas commented:
Rebase + last failing IT fixed.

Scheduled job that should completely pass now:
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/87/

cstamas commented:
This is going in, CI voted +0.5: linux passed okay, win is `0xdeadbeef`:
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/87/jdk=java-6x,label=win/console
